### DISCUSSION TITLE
New Tab Box & More

### DISCUSSION THREAD

**Discussion by madprops** [madprops/grasshopper#70] [Announcements]
The Tab Box has been worked on, it now uses the same item render function as the normal items, so it should have all functionality that normal items have.

Also, it now has more modes. It now has these modes:

1) Recent
2) Pins
3) Colors
4) Playing
5) Headers

Also, in `Themes` settings at the bottom there's a new setting to change the opacity of the background.
This can allow for interesting theme choices. The range can be from 0 (completely transparent) to 100 (no transparency. The default is 92.

There's a new `Previous Filter` command that goes restores the last used filter.

This is on `v2144`.

**Comment by madprops** [comment]
I removed some settings and effects on the tab box as I deemed them unnecessary. This includes the selected effect, as I want it to work more like a jumper launcher, instead of polluting with that feedback.

**Comment by N3C2L** [comment]
Thanks for the update, tabs in tab box is now hover button able + tags, cool!

But what is a jumper launcher?

**Reply by madprops** [reply]
I just meant like something you click to focus a tab, without having an active effect ... though I might give them an active effect, i'll see.

**Reply by N3C2L** [reply]
I would appreciate at least the underline option for that active tab, because then it doesn't confuse me that much with the highlighted active tab background in the tab list above.

**Reply by madprops** [reply]
Added `tab_box_active_effect`, `tab_box_icons`, `tab_box_colors`, `tab_box_tab_colors`, and `tab_box_taglist`.

**Reply by N3C2L** [reply]
Much obliged!

**Reply by N3C2L** [reply]
found an issue: the active tab (recent) in the tab box & active tab in the tab list are highlighted background & underlined (I have set it to underlined in tab box & highlighted background in the tab list) at the same time (they should show only one effect, not both together). this is causing it: if I set a custom text color & custom text background for the active tab (in the color settings).

**Reply by madprops** [reply]
What I'm seeing is that the tab box item gets colored the same as the normal item. And only the underline effect is applied. If you don't want the tab colors you can disable it in the tab box Show settings.

Oh wait... I found an issue. It's gonna be fixed in next version.

**Comment by madprops** [comment]
Added `Tab Box Auto Playing` setting. 
When this is enabled, the tab box will auto show when a tab starts playing.
It auto hides when no tabs are playing anymore.

**Comment by N3C2L** [comment]
Thank you, it works. 
I have also thought about an auto-hide feature for the tab box. When the cursor touches the Tab Box interface, the tab box automatically enlarges to a different size. When the cursor leaves the Tab Box interface, its previous size is restored.

**Comment by madprops** [comment]
I'm removing the auto hide portion of the setting, it will only show.

Also I'm adding an icon on the bottom left (footer), to toggle to Tab Box.

Also noted on that auto-resize idea.

Ok im adding the grow setting, with an animation even!

It's a menu where you select between normal, big, and huge, for the size to grow to.
Then it uses the previous size on mouseleave.

**Reply by N3C2L** [reply]
Thank you, that turned out well. The button in the footer is also great; it's helpful.

**Reply by madprops** [reply]
I'm going to upload a big update to the tab box soon. But it's mostly internal stuff.

**Reply by N3C2L** [reply]
Thanks, but now the Tab Box is not appearing when enabled. I think it's because of the footer. When the footer is disabled, the Tab Box won't show up when enabled.

**Reply by madprops** [reply]
Fixed now.

**Reply by N3C2L** [reply]
Thanks!

**Comment by N3C2L** [comment]
Under Settings > More > Max Recent Tabs. The value here does not seem to influence the value of Recent Tabs in the Tab Box, as far as I can see. But if I'm not mistaken, there used to be such a value for setting the number of Recent Tabs in the Tab Box?

I have the color red for the titles in my unloaded tabs. It would be nice if this text color for unloaded tabs is also visible in the tab box. Because then I would have a better overview of which tabs I have really used or not used recently.

**Reply by madprops** [reply]
What do you mean, I see red text color for unloaded in both the items and the tab box (if I set it up). What other tab colors do you have enabled?

**Reply by madprops** [reply]
I'm moving Unloaded and Loaded to be higher priority than Pinned and Normal

**Reply by madprops** [reply]
And yeah you're right. Max Recent Tabs controls the items in the Recent Tabs when right clicking Step Back. There used to be an active history setting but I removed that. I added `Tab Box Max` in zones to control the tab box. Upload soon.

**Reply by madprops** [reply]
Tab Box got big enough. Moving the settings to their own category called Tab Box.

**Comment by N3C2L** [comment]
Thank you. I apologize; I seem to have overlooked that "Tab Box Tab Colors" also sets the value for the title. Yes, it works; the unloaded title tabs are red. However, there is now the following issue:

1. Activate "Tab Box Tab Colors"
2. Change the "Tab Box Active Effect" to underline
3. Change the "Selected Effect" in General to background
4. Go to Colors Settings > check active tabs (background), for instance to green
Now in the Tab Box both are visible: Underline + Background
If a "Tab Box Active Effect" is set by a user, it should override the "active tabs (background)" in Colors Settings, so that not both effects are visible in the tab box.

Very nice that there is now a separate Tab Box category.
But would it also be possible for the loaded tabs to be sorted at the top and the unloaded tabs below? For example, when selecting an unloaded tab in the Tab Box under "Colors," it moves up to the other loaded tabs. This way, the loaded tabs are grouped together at the top of the Tab Box, and the unloaded tabs are below. That would be a amazing. But this should only be visible in the Tab Box. The tabs should not be moved in the tab list above the Tab Box.

**Reply by madprops** [reply]
> Now in the Tab Box both are visible: Underline + Background

You mean a green background + underline?
The green background is the tab color, is not an effect

**Reply by N3C2L** [reply]
Hm. But under > General > Selected Effect > you can choose between background or underline. So that only one of them is visible, not both like in the problem I described above.

**Reply by madprops** [reply]
When I tried to do what you described, I just saw a green background (as set in Colors), and the underline effect, on the active tab

**Reply by madprops** [reply]
![activ3234](https://github.com/madprops/grasshopper/assets/206498/24c74748-5a6e-4d92-a04d-7e7960b38acc)

**Reply by madprops** [reply]
This is with Selected Effect as Background in General.

**Reply by madprops** [reply]
I think you're confusing what background effect is. If a tab is set to have a green background, that's the background it will have, overriding the background effect, but maybe that can be improved, i'll try making the background effect to affect custom backgrounds as well, give it some contrast somehow

**Reply by N3C2L** [reply]
Yeah but only one effect is visible in the tab list above when select "background" as the selected effect in general. But both effects are visible in the tab box. What if a users like me wants only the underline effect as the selected effect in the tab box. There should be an option then to disable the active tab custom color in the tab box.

**Reply by madprops** [reply]
Can you screenshot what you mean? I want to see the difference between the item container and the tab box

**Reply by madprops** [reply]
If you mean that both are green, that's just the tab color, because you enabled tab colors in the tab box, it's not an additional effect, you specified that active tabs are green. That said, im working on giving "background" effect visible on all colored tabs.

**Reply by N3C2L** [reply]
If you have set no custom background for the active tab in the color settings, then only one effect is displayed when changing the setting under general > select effect. But it's confusing, when you have set a custom color for the active tab, and then change the select effect for the active tab in general (for example underline), then both effects are displayed. I think only one effect should be visible independently to what custom active color is given to a tab. At least in the tab box. That's why I suggested an option then to disable the active tab custom color in the tab box.

**Reply by madprops** [reply]
You can disable the colors from the tab box. The green is not the effect, it's the color of that kind of tab. The background effect in theory is a slight contrast of whatever background color it has, slightly darker, or slightly lighter (which im improving right now).

**Reply by N3C2L** [reply]
But if I disable them, I wouldn't see the unloaded/loaded text color anymore in the Tab Box.

**Reply by madprops** [reply]
Idk. I'm going to upload a color update soon, maybe it's better now that all colors have a contrast.

**Reply by N3C2L** [reply]
I think this could solve my problem: If a user can choose which colors, background or text, should be enabled in the tab box. There could be additionally checkboxes under each tab in the color category, for enabling the color in the tab box or not.
But if this is too much than my previous suggestion would solve also my problem: I suggested an option then to disable the active tab custom background color in the tab box (but leave all other tab colors intact)

**Reply by madprops** [reply]
No that is making the already complex color system even more complex

**Reply by madprops** [reply]
What if a pinned tab is active, and the pinned tab is purple because Pinned color is purple
I'd have to add a checkbox to every color setting

**Reply by madprops** [reply]
I'll experiment adding a menu instead of a checkbox

**Comment by madprops** [comment]
Alright it works. But all users will have to set it up since it defaults to `none`.
Modes: none, items, tab_box, everywhere
Upload in a while

It required a re-write of the color function. But it should be more granular now.

**Reply by N3C2L** [reply]
Thank you, looking forward to being able to try it out.

**Comment by madprops** [comment]
Alright it's up. Modes are: None, Normal, Tab Box, and Everywhere. Since it was a big change there might be bugs.

**Reply by N3C2L** [reply]
Excellent, thanks again for this update! Everything works smoothly, no bug found yet.

**Comment by madprops** [comment]
The tab box now has an independent color mode (for green, red, etc) 
So you can tell it to use only icons or background etc.

**Comment by N3C2L** [comment]
Thank you, the new independent color mode for the Tab Box is valuable.

It would be wonderful if the "All Tags" + "All Icons" filtering in the Tab Box could be introduced as separate views. This way, I could view other perspectives with tags or other to-do icons independently of my colored tabs (by the way, thanks for the icons introduced in the Edit Icon Cmd Menu). And these tabs would then be readily available to me in the "All Icons" and "All Tags" filtering in the Tab Box.

Now that you can actively click on the headers, I would prefer if these headers could appear in the Tab Box in the other sections (Colors, Pins, Recent, All Tags, All Icons). Perhaps an option for this would be ideal, as not every user may want to see these headers there. But I think users like me, who actively click on the headers to gain orientation, would like to see these header tabs in the Tab Box in the other sections as well.

**Comment by madprops** [comment]
Alright I'm working on this. It made me refactor the tab box code, it's cleaner now.

**Comment by N3C2L** [comment]
Thank you for this implementation. It is very pleasant to use.
